src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -324,6 +326,25 @@ uct_cuda_copy_md_query_attributes(uct_cuda_copy_md_t *md, const void *address,
          * allocated in a context should also allows us to identify
          * virtual/stream-ordered CUDA allocations. */
         mem_info->type = UCS_MEMORY_TYPE_CUDA_MANAGED;
+
+        cu_err =
+            cuMemRangeGetAttribute((void*)&pref_loc, 4,
+                                   CU_MEM_RANGE_ATTRIBUTE_PREFERRED_LOCATION,
+                                   (CUdeviceptr)address, length);
+        if ((cu_err == CUDA_SUCCESS) && (pref_loc != CU_DEVICE_INVALID)) {
+            if (pref_loc == CU_DEVICE_CPU) {
+                mem_info->sys_dev = UCS_SYS_DEVICE_ID_UNKNOWN;
+            } else {
+                /* save original sys_dev before overwriting, restore if unable
+                 * to find preferred location that user has set */
+                orig_sys_dev = mem_info->sys_dev;
```

Comment:
maybe we should restructure this - first check if the memory is managed, and try to use PREFERRED_LOCATION to get the system devices. If the memory is not managed, or if the PREFERRED_LOCATION check fails - fallback to calling uct_cuda_base_get_sys_dev

Reply:
@yosefe Agree with the restructuring other than for the case when memory is managed, and preferred location attribute is set but the location set by the user does not correspond to a device that ucs_sys_topo [recognizes](https://github.com/openucx/ucx/pull/9011/files/221bcb57c60ede56b78567afe18458ab6a5711bf#diff-f990528a14758f5a40bc106bd5fb5888513d667e4fc0b80256fe09dee83015c0R344). In this case, we'd have to query the device associated with cuda_context again using cuptrgetattributes call, and then query the sys_device for it. Is it ok to leave this logic as is? After all, we're trying to avoid the use of the extra variables orig_sys_device and pref_location right?

Reply:
It would be unexpected that preferred location is an unrecognized device.
In this cat I'd print ucs_diag() message and return UNKNOWN device.